lib/husky_musher/utils/redcap.py

```diff
@@ -0,0 +1,535 @@
+        'redcap_repeat_instance': str(get_todays_repeat_instance()),
+        'testing_trigger': YES,
+        'testing_type': KIOSK_WALK_IN,
+        'testing_date': str(datetime.today().date()),
+        'testing_determination_internal_complete': COMPLETE,
+    }]
+
+    data = {
+        'token': REDCAP_API_TOKEN,
+        'content': 'record',
+        'format': 'json',
+        'type': 'flat',
+        'overwriteBehavior': 'normal',
+        'forceAutoNumber': 'false',
+        'data': json.dumps(record),
+        'returnContent': 'ids',
+        'returnFormat': 'json'
+    }
+
+    response = requests.post(REDCAP_API_URL, data=data)
+    response.raise_for_status()
+
+    assert len(response.json()) == 1, \
+        f"REDCap updated {len(response.json())} records, expected 1."
+
+
+def need_to_create_new_td_for_today(instances: Dict[str, int]) -> bool:
+    """
+    Returns True if we need to create a new TD instance for today. Otherwise,
+    returns False.
+
+    We need to create a new TD instance for today in the following conditions:
+        1. No TD instance with [testing_trigger] = "Yes" exists in the past 7
+           days.
+        2. No complete TOS instance exists on or after the target TD instance,
+           and a complete KR instance exists on or after the target TD instance.
+        3. A complete TOS instance exists on or after the target TD instance,
+           the TOS instance is not from today, and a complete KR instance exists
+           on or after the target TD instance.
+
+    *target_instance* is a TD instance number with [testing_trigger] = "Yes" in the
+    past 7 days.
+
+    *complete_tos_instance* is a TOS instance number marked complete on or after the
+    given *target_instance*.
+
+    *complete_kr_instance* is a KR instance number marked complete on or after the
+    given *target_intance*.
+
+    >>> need_to_create_new_td_for_today({'target': None, 'complete_tos': 1, 'complete_kr': 1})
+    True
+
+    >>> need_to_create_new_td_for_today({'target': 1, 'complete_tos': None, 'complete_kr': 1})
+    True
+
+    >>> need_to_create_new_td_for_today({'target': 1, 'complete_tos': 1, 'complete_kr': 1})
+    True
+
+    >>> need_to_create_new_td_for_today({'target': 1, 'complete_tos': None, 'complete_kr': None})
+    False
+
+    >>> need_to_create_new_td_for_today({'target': 1, 'complete_tos': None, 'complete_kr': 1})
+    True
+
+    >>> need_to_create_new_td_for_today({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': 1})
+    False
+
+    >>> need_to_create_new_td_for_today({'target': 1, \
+        'complete_tos': get_todays_repeat_instance(), 'complete_kr': None})
+    False
+    """
+    if not instances['target']:
+        return True
+
+    complete_tos_instance = instances['complete_tos']
+    if complete_tos_instance is None \
+        or (complete_tos_instance is not None \
+            and complete_tos_instance != get_todays_repeat_instance()):
```

Comment:
I was trying to parse this conditional, and in the process realized that (I think) it reduces to just:

```
if instances['complete_tos'] != get_todays_repeat_instance():
```

Reply:
Yep, good catch. That's much easier to read.

Reply:
I love that I can just run the doctests and unit tests to make sure nothing broke.